README.md
# R Packages: sinib
Sum of Independent Non-Identical Binomial Random Variables

## Description
Density function, distribution function, quantile function and random number generation for the sum of independent non-identical	binomial distribution with parameters `size` and `prob`.
 
## Installation
In R console, type 
```
install.packages('sinib')
library('sinib')
```

## Documentation
The documentation for the four main functions `dsinib`, `psinib`, `qsinib`, and `rsinib` can be assessed in R by typing
```
?dsinib
?psinib
?qsinib
?rsinib
```

## Example
Suppose `S=X1+X2+X3` where `X1~binom(12,0.074)`, `X2~binom(14,0.039)` and `X3~binom(4,0.095)`. If you would like to calculate the probability `P(S=3)` and `P(S=20)`, you would do the following.

```
size=as.integer(c(12, 14, 4))
prob=c(0.074, 0.039, 0.095)
x=c(3,20)
dsinib(x,size,prob)
# [1] 1.675016e-01 3.167624e-18
```

## Citation 
Please cite Boxiang Liu and Thomas Quertermous (2017) Approximating the Sum of Independent Non-Identical Binomial Random Variables. arXiv.

NEWS.md
# sinib 1.0.0
First major release. 

# sinib 1.0.1
bugs fixes: 
1. Fixed numerical instability issues.
2. Removed input integer constraint. 

R/test.R
# Small size, small probability:
size=as.integer(c(12, 14, 4, 2, 20, 17, 11, 1, 8, 11))
prob=c(0.074, 0.039, 0.095, 0.039, 0.053, 0.043, 0.067, 0.018, 0.099, 0.045)

x=as.integer(seq(1,19,2))
true=c(0.0164, 0.0994, 0.1716, 0.1346, 0.0587, 0.0160, 0.002913, 0.0003752, 0.00003544, 0.000002525)
stopifnot(all(((dsinib(x,size,prob)-true))/true<0.001))

true=c(0.9972, 0.9308, 0.6847, 0.3481, 0.1187, 0.0276, 0.004546, 0.0005435, 0.00004855, 0.000003313)
q=as.integer(seq(0,18,2))
stopifnot(all((psinib(q,size,prob,lower.tail=FALSE)-true)/true<0.01))

p=psinib(q,size,prob)
stopifnot(all(qsinib(p,size,prob)==q))


# Large size, large probability:
size=as.integer(c(120, 140, 40, 20, 200, 170, 110, 10, 80, 110))
prob=c(0.74, 0.39, 0.95, 0.39, 0.53, 0.43, 0.67, 0.18, 0.99, 0.45)

x=as.integer(seq(510,640,10))
true=c(0.000002363, 0.00003730, 0.0003638, 0.002195, 0.008202, 0.0190, 0.0272, 0.0242, 0.0133, 0.004501, 0.0009419, 0.0001213, 0.000009581, 0.0000004630)
stopifnot(all(((dsinib(x,size,prob)-true))/true<0.001))

true=c(0.99993616, 0.9998791, 0.998549, 0.9889, 0.9444, 0.8161, 0.5825, 0.3140, 0.1194, 0.0306, 0.005133, 0.0005522, 0.00003757, 0.000001599)
q=as.integer(seq(509,639,10))
stopifnot(all((psinib(q,size,prob,lower.tail=FALSE)-true)/true<0.01))

p=psinib(q,size,prob)
stopifnot(all(qsinib(p,size,prob)==q))


tests/test.R
library(sinib)

# Small size, small probability:
size=c(12, 14, 4, 2, 20, 17, 11, 1, 8, 11)
prob=c(0.074, 0.039, 0.095, 0.039, 0.053, 0.043, 0.067, 0.018, 0.099, 0.045)

x=seq(1,19,2)
true=c(0.0164, 0.0994, 0.1716, 0.1346, 0.0587, 0.0160, 0.002913, 0.0003752, 0.00003544, 0.000002525)
stopifnot(all(((dsinib(x,size,prob)-true))/true<0.001))

true=c(0.9972, 0.9308, 0.6847, 0.3481, 0.1187, 0.0276, 0.004546, 0.0005435, 0.00004855, 0.000003313)
q=seq(0,18,2)
stopifnot(all((psinib(q,size,prob,lower.tail=FALSE)-true)/true<0.01))

p=psinib(q,size,prob)
stopifnot(all(qsinib(p,size,prob)==q))


# Large size, large probability:
size=c(120, 140, 40, 20, 200, 170, 110, 10, 80, 110)
prob=c(0.74, 0.39, 0.95, 0.39, 0.53, 0.43, 0.67, 0.18, 0.99, 0.45)

x=seq(510,640,10)
true=c(0.000002363, 0.00003730, 0.0003638, 0.002195, 0.008202, 0.0190, 0.0272, 0.0242, 0.0133, 0.004501, 0.0009419, 0.0001213, 0.000009581, 0.0000004630)
stopifnot(all(((dsinib(x,size,prob)-true))/true<0.001))

true=c(0.99993616, 0.9998791, 0.998549, 0.9889, 0.9444, 0.8161, 0.5825, 0.3140, 0.1194, 0.0306, 0.005133, 0.0005522, 0.00003757, 0.000001599)
q=seq(509,639,10)
stopifnot(all((psinib(q,size,prob,lower.tail=FALSE)-true)/true<0.01))

p=psinib(q,size,prob)
stopifnot(all(qsinib(p,size,prob)==q))

R/sinib.R
source('R/utils.R')
#' @importFrom stats uniroot
#' @importFrom stats dnorm
#' @importFrom stats pnorm
#' @importFrom stats rbinom
#' @importFrom stats runif
NULL


#' Distribution of Sum of Independent Non-Identical Binomial Random Variables
#' 
#' Density function, distribution function, quantile function, and random number generation for the sum of independent non-identical binomial random variables 
#' 
#' Suppose S is a random variable formed by summing R independent non-identical random variables \eqn{X_r}{Xr}, \eqn{r = 1,...,R}. \deqn{S = \sum_{r=1}^R X_r}{S = X1+X2+...XR}
#' \code{size} and \code{prob} should both be vectors of length R. The first elements of \code{size} and \code{prob} specifies \eqn{X_1}{X1}, the second elements specifies \eqn{X_2}{X2}, so on and so forth. The probability \eqn{F(S)} is calculated using Daniels' second-order continuity-corrected saddlepoint approximation. The density \eqn{p(S)} is calculated using second-order saddlepoint mass approximation with Butler's normalization. 
#' 
#' @param x,q integer vector of quantiles.
#' @param p numeric vector of probabilities. 
#' @param n numeric scalar to indicate number of observations.
#' @param size integer vector of number of trials (see detail).
#' @param prob numeric vector of success probabilities (see detail).
#' @param lower.tail logical; if TRUE, probabilities are \eqn{P[S<=s]}, otherwise, \eqn{P[S>s]}.
#' @param log,log.p logical; if TRUE, probabilities p are given as log(p).
#' @return qsinib gives the cumulative distribution of sum of independent non-identical random variables. 
#' @source See Eisinga et al (2012) Saddlepoint approximations for the sum of independent non-identically distributed binomial random variables. Available from \url{http://onlinelibrary.wiley.com/doi/10.1111/stan.12002/full}
#' 
#' @examples
#' # Calculating the density and probability:
#' size <- c(12, 14, 4, 2, 20, 17, 11, 1, 8, 11)
#' prob <- c(0.074, 0.039, 0.095, 0.039, 0.053, 0.043, 0.067, 0.018, 0.099, 0.045)
#' q <- x <- seq(1, 19, 2)
#' dsinib(x, size, prob)
#' psinib(q, size, prob)
#' 
#' # Generating random samples:
#' rsinib(100, size, prob)
#' 
#' # Calculating quantiles:
#' p <- psinib(q, size, prob) 
#' qsinib(p, size, prob)
#' 
#' @export
psinib=function(q,size,prob,lower.tail=TRUE,log.p=FALSE){
	if (!all(size == as.integer(size))){
		warning('size should be integers.')
	}
	if (!all(q == as.integer(q))){
		warning('q should be integers.')
	}
	
	n=as.integer(size)
	p=prob
	s=as.integer(q)

	mu=sum(n*p)

	w=function(u,n,p){
		K_=K(u,n,p)
		Kp_=Kp(u,n,p)
		return(sign(u)*sqrt(2*u*Kp_-2*K_))
	}
	u1=function(u,n,p){
		Kpp_=Kpp(u,n,p)
		return((1-exp(-u))*sqrt(Kpp_))
	}
	p3=function(u,n,p,s,mu){
		w_=w(u,n,p)
		u1_=u1(u,n,p)
		if (u1_==0 | s==mu) {
			Kpp0=sum(n*p*(1-p))
			Kppp0=sum(n*p*(1-p)*(1-2*p))
			return(1/2-(2*pi)^(-1/2)*((1/6)*Kppp0*Kpp0^(-3/2)-(1/2)*Kpp0^(-1/2)))
		} else {
			return(1-pnorm(w_)-dnorm(w_)*(1/w_-1/u1_))
		}
	}
	
	
	u2=function(u,n,p){
		return(u*sqrt(Kpp(u,n,p)))
	}
	k3=function(u,n,p){
		return(Kppp(u,n,p)*Kpp(u,n,p)^(-3/2))
	}
	k4=function(u,n,p){
		return(Kpppp(u,n,p)*Kpp(u,n,p)^(-2))
	}
	p4=function(u,n,p,s,mu){
		u2_=u2(u,n,p)
		k3_=k3(u,n,p)
		k4_=k4(u,n,p)
		w_=w(u,n,p)
		p3_=p3(u,n,p,s,mu)
		u1_=u1(u,n,p)
		if (u1_==0 | s==mu){
			return(p3_)
		} else {
			return(p3_-dnorm(w_)*((1/u2_)*((1/8)*k4_-(5/24)*k3_^2)-1/(u2_^3)-k3_/(2*u2_^2)+1/w_^3))
		}
		
	}
	
	calc_p4=function(s){
		if (s == sum(n)){
			p4_=prod(p^n)
		} else {
			u_hat=tryCatch({uniroot(saddlepoint,lower=-100,upper=100,tol = 0.0001,n=n,p=p,s=s)$root},error=function(e){return(NA)})
			if (is.na(u_hat)){
				n_trial=100000
				n_binom=length(p)
				set.seed(42)
				mat=matrix(rbinom(n_trial*n_binom,n,p),nrow=n_binom,ncol=n_trial)
				S=colSums(mat)
				p4_=sum(S>=s)/length(S)
			} else {
				p4_=p4(u_hat,n,p,s,mu)
			}
		}
		return(p4_)
	}
	

	p4_=sapply(s+1,calc_p4)

	if (lower.tail){
		if (log.p){
			return(log(1-p4_))
		} else {
			return(1-p4_)	
		}
	} else {
		if (log.p){
			return(log(p4_))
		} else {
			return(p4_)
		}
			
	}
}

#' @rdname psinib
#' @export
dsinib=function(x,size,prob,log=FALSE){
	if (!all(size == as.integer(size))){
		warning('size should be integers.')
	}
	if (!all(x == as.integer(x))){
		warning('x should be integers.')
	}
	n=as.integer(size)
	p=prob
	s=as.integer(x)
	
	p1=function(u,n,p,s){
		return((2*pi*Kpp(u,n,p))^(-1/2)*exp(K(u,n,p)-u*s))
	}
	p2=function(u,n,p,s){
		return(p1(u,n,p,s)*(1+(1/8)*Kpppp(u,n,p)/Kpp(u,n,p)^2-(5/24)*Kppp(u,n,p)^2/Kpp(u,n,p)^3))
	}
	
	calc_p2=function(s){
		u_hat=tryCatch({uniroot(saddlepoint,lower=-100,upper=100,tol = 0.0001,n=n,p=p,s=s)$root},error=function(e){return(NA)})
		p2_=p2(u_hat,n,p,s)
		return(p2_)
	}
	p2_=sapply(1:(sum(n)-1),calc_p2)
	
	p2b=function(n,p,p2_){
		p0=prod((1-p)^n)
		pmx=prod(p^n)
		p2b_=rep(NA,sum(n)+1)
		p2b_[1]=p0
		p2b_[length(p2b_)]=pmx
		p2b_[2:sum(n)]=(1-p0-pmx)*p2_/sum(p2_)
		return(p2b_)
	}
	p2b_=p2b(n,p,p2_)
	
	if (log){
		return(log(p2b_[s+1]))
	} else {
		return(p2b_[s+1])
	}
		
}

#' @rdname psinib
#' @export
rsinib=function(n,size,prob){
	cdf=psinib(q=0:sum(size),size = size,prob = prob)
	x=runif(n,0,1)
	y=sapply(x,function(x) {min(which(cdf>=x))-1})
	return(y)
}

#' @rdname psinib
#' @export
qsinib=function(p,size,prob){
	cdf=psinib(q=0:sum(size),size = size,prob = prob)
	y=sapply(p,function(x) {min(which(cdf>=x))-1})
	return(y)
}

manuscript/scripts/two_binomials.R
library(foreach)
library(data.table)
library(cowplot)
library(sinib)

# Comparison of CDF between truth and approximation:
data=foreach(m=c(10,100,1000),.combine='rbind')%do%{
	foreach(n=c(10,100,1000),.combine='rbind')%do%{
		foreach(p=c(0.1, 0.5, 0.9),.combine='rbind')%do%{
			a=pbinom(q=0:(m+n),size=(m+n),prob = p)
			b=psinib(q=0:(m+n),size=as.integer(c(m,n)),prob=c(p,p))
			data.table(s=seq_along(a),truth=a,approx=b,m=m,n=n,p=p)
		}
	}
}

p=ggplot(data,aes(x=truth,y=approx,color=as.character(p)))+geom_point(alpha=0.5)+facet_grid(m~n)+theme_bw()+scale_color_discrete(name='prob')+xlab('Truth')+ylab('Approximation')

save_plot('~/Documents/tools/sinib/manuscript/figures/truth_vs_approximation.pdf',p,base_width=6,base_height=6)


p2=ggplot(data[m==100&n==100],aes(x=s,y=truth-approx,color=as.character(p)))+geom_point(alpha=0.5)+theme_bw()+scale_color_discrete(name='prob')+xlab('Quantile')+ylab('Truth-Approximation')+geom_vline(xintercept=200*0.5,color='green',linetype='longdash')+geom_vline(xintercept=200*0.1,color='red',linetype='longdash')+geom_vline(xintercept=200*0.9,color='blue',linetype='longdash')

save_plot('~/Documents/tools/sinib/manuscript/figures/truth_minus_approximation.pdf',p2,base_width=6,base_height=4)


# Comparison of PDF between truth and approximation:
data=foreach(m=c(10,100,1000),.combine='rbind')%do%{
	foreach(n=c(10,100,1000),.combine='rbind')%do%{
		foreach(p=c(0.1, 0.5, 0.9),.combine='rbind')%do%{
			a=dbinom(x=0:(m+n),size=(m+n),prob = p)
			b=dsinib(x=0:(m+n),size=as.integer(c(m,n)),prob=c(p,p))
			data.table(s=seq_along(a),truth=a,approx=b,m=m,n=n,p=p)
		}
	}
}

p3=ggplot(data,aes(x=truth,y=approx,color=as.character(p)))+geom_point(alpha=0.5)+facet_grid(m~n)+theme_bw()+scale_color_discrete(name='prob')+xlab('Truth')+ylab('Approximation')

save_plot('~/Documents/tools/sinib/manuscript/figures/truth_vs_approximation_pdf.pdf',p3,base_width=6,base_height=6)


p4=ggplot(data[m==100&n==100],aes(x=s,y=truth-approx,color=as.character(p)))+geom_point(alpha=0.5)+theme_bw()+scale_color_discrete(name='prob')+xlab('Quantile')+ylab('Truth-Approximation')+geom_vline(xintercept=200*0.5,color='green',linetype='longdash')+geom_vline(xintercept=200*0.1,color='red',linetype='longdash')+geom_vline(xintercept=200*0.9,color='blue',linetype='longdash')

save_plot('~/Documents/tools/sinib/manuscript/figures/truth_minus_approximation_pdf.pdf',p4,base_width=6,base_height=4)


manuscript/scripts/health_monitor.R
library(sinib)
library(foreach)
library(cowplot)
library(data.table)

# Sinib:
size=as.integer(c(12, 14, 4, 2, 20, 17, 11, 1, 8, 11))
prob=c(0.074, 0.039, 0.095, 0.039, 0.053, 0.043, 0.067, 0.018, 0.099, 0.045)
approx=dsinib(0:sum(size),size,prob)
approx=data.frame(s=0:sum(size),pdf=approx,type='saddlepoint')

# Simulation:
data=foreach(n_sim=10^c(3:6,8),.combine='rbind')%do%{
	ptm=proc.time()
	n_binom=length(prob)
	set.seed(42)
	mat=matrix(rbinom(n_sim*n_binom,size,prob),nrow=n_binom,ncol=n_sim)
	
	S=colSums(mat)
	sim=sapply(X = 0:sum(size), FUN = function(x) {sum(S==x)/length(S)})
	
	print(proc.time()-ptm)
	
	data.table(s=0:sum(size),pdf=sim,type=n_sim)
	
}


	

data=rbind(data,approx)
truth=data[type=='1e+08',]

merged=merge(truth[,list(s,pdf)],data,by='s',suffixes=c('_truth','_approx'))
merged=merged[type!='1e+08',]

p=ggplot(merged,aes(pdf_truth,pdf_approx))+geom_point()+facet_grid(~type)+geom_abline(intercept=0,slope=1)+theme_bw()+xlab('Truth')+ylab('Approx')
save_plot('/Users/boshliu/Documents/tools/sinib/manuscript/figures/health_monitor_truth_vs_approx_pdf.pdf',p,base_height=3,base_width=8)

merged[,diff:=pdf_truth-pdf_approx]

p2=ggplot(merged,aes(s,diff))+geom_point()+facet_grid(~type)+theme_bw()+xlab('Quantile')+ylab('Truth-Approx')
save_plot('/Users/boshliu/Documents/tools/sinib/manuscript/figures/health_monitor_truth_minus_approx_pdf.pdf',p2,base_height=3,base_width=8)



# time: 
ptm=proc.time()
n_binom=length(prob)
mat=matrix(rbinom(n_sim*n_binom,size,prob),nrow=n_binom,ncol=n_sim)
S=colSums(mat)
sim=sapply(X = 0:sum(size), FUN = function(x) {sum(S==x)/length(S)})
proc.time()-ptm

ptm=proc.time()
approx=dsinib(0:sum(size),size,prob)
proc.time()-ptm


R/liu.R
library(foreach)
library(data.table)
library(cowplot)
library(sinib)

#-----------#
# Functions #
#-----------#
p_norm_app = function(q,size,prob){
	mu = sum(size*prob)
	sigma = sqrt(sum(size*prob*(1-prob)))
	pnorm(q, mean = mu, sd = sigma)
}

d_norm_app = function(x,size,prob){
	mu = sum(size*prob)
	sigma = sqrt(sum(size*prob*(1-prob)))
	dnorm(x, mean = mu, sd = sigma)
}

#----------#
# Figure 1 #
#----------#
# Comparison of CDF between truth and approximation:
data=foreach(m=c(10,100,1000),.combine='rbind')%do%{
	foreach(n=c(10,100,1000),.combine='rbind')%do%{
		foreach(p=c(0.1, 0.5, 0.9),.combine='rbind')%do%{
			a=pbinom(q=0:(m+n),size=(m+n),prob = p)
			b=psinib(q=0:(m+n),size=c(m,n),prob=c(p,p))
			c=p_norm_app(q=0:(m+n),size=c(m,n),prob = c(p,p))
			data.table(s=seq_along(a),truth=a,saddle=b,norm=c,m=m,n=n,p=p)
		}
	}
}
data[,m:=paste0('m = ',m)]
data[,p:=paste0('p = ',p)]
data = melt(data,measure.vars = c('saddle','norm'))

for (i in c(10,100,1000)){
	p=ggplot(data[n==i],aes(x=truth,y=value,color=p,linetype=variable))+
		geom_line()+
		facet_grid(p~m)+
		theme_bw()+
		scale_color_discrete(name='prob',guide = 'none')+
		xlab('Truth')+
		ylab('Approximation') + 
		scale_linetype_discrete(name = '', breaks = c('saddle','norm'), labels = c('Saddlepoint','Gassian')) + 
		theme(legend.position = 'top')
	save_plot(sprintf('~/Documents/tools/sinib/manuscript/figures/truth_vs_approximation_n%s.pdf',i),p,base_width=6,base_height=6)
}


#----------#
# Figure 2 #
#----------#
data=foreach(m=c(100),.combine='rbind')%do%{
	foreach(n=c(100),.combine='rbind')%do%{
		foreach(p=c(0.1, 0.5, 0.9),.combine='rbind')%do%{
			a=pbinom(q=0:(m+n),size=(m+n),prob = p)
			b=psinib(q=0:(m+n),size=as.integer(c(m,n)),prob=c(p,p))
			c=p_norm_app(q=0:(m+n),size=c(m,n),prob = c(p,p))
			data.table(s=seq_along(a),truth=a,saddle=b,norm=c,m=m,n=n,p=p)
		}
	}
}
data = melt(data,measure.vars = c('saddle','norm'),value.name = 'approx', variable.name = 'Method')
data[,`Relative error` := (truth-approx)/truth]
data[,Error := (truth-approx)]
data[,p:=paste0('p = ',p)]
data = melt(data,measure.vars = c('Relative error','Error'),value.name = 'error', variable.name = 'type')

p2=ggplot(data[Method == 'saddle'],aes(x=s,y=error,color=p))+
	geom_point(alpha=0.5)+theme_bw()+
	facet_wrap(type~p,scales='free')+
	xlab('Quantile')+
	ylab(expression('Truth - Approximation'~~~~~~~~~~~~~~~frac(Truth - Approximation,Truth)))+
	scale_color_discrete(guide='none') + 
	geom_vline(xintercept=200*0.5,color='green',linetype='longdash')+
	geom_vline(xintercept=200*0.1,color='red',linetype='longdash')+
	geom_vline(xintercept=200*0.9,color='blue',linetype='longdash')
	
save_plot('~/Documents/tools/sinib/manuscript/figures/truth_minus_saddlepoint_approximation.pdf',p2,base_width=6,base_height=4.5)


p2=ggplot(data[Method == 'norm'],aes(x=s,y=error,color=p))+
	geom_point(alpha=0.5)+theme_bw()+
	facet_wrap(type~p,scales='free')+
	xlab('Quantile')+
	ylab(expression('Truth - Approximation'~~~~~~~~~~~~~~~frac(Truth - Approximation,Truth)))+
	scale_color_discrete(guide='none') + 
	geom_vline(xintercept=200*0.5,color='green',linetype='longdash')+
	geom_vline(xintercept=200*0.1,color='red',linetype='longdash')+
	geom_vline(xintercept=200*0.9,color='blue',linetype='longdash')

save_plot('~/Documents/tools/sinib/manuscript/figures/truth_minus_gaussian_approximation.pdf',p2,base_width=6,base_height=4.5)

#----------#
# Figure 3 #
#----------#
# Comparison of PDF between truth and approximation:
data=foreach(m=c(10,100,1000),.combine='rbind')%do%{
	foreach(n=c(10,100,1000),.combine='rbind')%do%{
		foreach(p=c(0.1, 0.5, 0.9),.combine='rbind')%do%{
			a=dbinom(x=0:(m+n),size=(m+n),prob = p)
			b=dsinib(x=0:(m+n),size=as.integer(c(m,n)),prob=c(p,p))
			c=d_norm_app(x=0:(m+n),size=c(m,n),prob=c(p,p))
			data.table(s=seq_along(a),truth=a,saddle=b,norm=c,m=m,n=n,p=p)
		}
	}
}

data[,m:=paste0('m = ',m)]
data[,p:=paste0('p = ',p)]
data = melt(data,measure.vars = c('saddle','norm'))

for (i in c(10,100,1000)){
	p3=ggplot(data[n==i],aes(x=truth,y=value,color=p,linetype=variable))+
		geom_line()+
		facet_wrap(m~p,scales='free')+
		theme_bw()+
		scale_color_discrete(guide = 'none')+
		xlab('Truth')+
		ylab('Approximation') + 
		scale_linetype_discrete(name = '', breaks = c('saddle','norm'), labels = c('Saddlepoint','Gassian')) + 
		theme(legend.position = 'top')
	save_plot(sprintf('~/Documents/tools/sinib/manuscript/figures/truth_vs_approximation_pdf_n%s.pdf',i),p3,base_width=6,base_height=6)
}


#----------# 
# Figure 4 #
#----------#
data=foreach(m=c(100),.combine='rbind')%do%{
	foreach(n=c(100),.combine='rbind')%do%{
		foreach(p=c(0.1, 0.5, 0.9),.combine='rbind')%do%{
			a=dbinom(x=0:(m+n),size=(m+n),prob = p)
			b=dsinib(x=0:(m+n),size=as.integer(c(m,n)),prob=c(p,p))
			c=d_norm_app(x=0:(m+n),size=c(m,n),prob = c(p,p))
			data.table(s=seq_along(a),truth=a,saddle=b,norm=c,m=m,n=n,p=p)
		}
	}
}
data = melt(data,measure.vars = c('saddle','norm'),value.name = 'approx', variable.name = 'Method')
data[,`Relative error` := (truth-approx)/truth]
data[,Error := (truth-approx)]
data[,p:=paste0('p = ',p)]
data = melt(data,measure.vars = c('Relative error','Error'),value.name = 'error', variable.name = 'type')

p4=ggplot(data[Method == 'saddle'],aes(x=s,y=error,color=p))+
	geom_point(alpha=0.5)+theme_bw()+
	facet_wrap(type~p,scales='free')+
	xlab('Quantile')+
	ylab(expression('Truth - Approximation'~~~~~~~~~~~~~~~frac(Truth - Approximation,Truth)))+
	scale_color_discrete(guide='none') + 
	geom_vline(xintercept=200*0.5,color='green',linetype='longdash')+
	geom_vline(xintercept=200*0.1,color='red',linetype='longdash')+
	geom_vline(xintercept=200*0.9,color='blue',linetype='longdash')

save_plot('~/Documents/tools/sinib/manuscript/figures/truth_minus_saddlepoint_approximation_pdf.pdf',p4,base_width=6,base_height=4.5)


#-----------------------#
# Healthcare monitoring #
#-----------------------#
library(sinib)
library(foreach)
library(cowplot)
library(data.table)

size=as.integer(c(12, 14, 4, 2, 20, 17, 11, 1, 8, 11))
prob=c(0.074, 0.039, 0.095, 0.039, 0.053, 0.043, 0.067, 0.018, 0.099, 0.045)

# Sinib:
approx=dsinib(0:sum(size),size,prob)
approx=data.frame(s=0:sum(size),pdf=approx,type='saddlepoint')

# Gauss: 
gauss_approx = d_norm_app(0:sum(size),size,prob)
gauss_approx = data.frame(s=0:sum(size),pdf=gauss_approx,type='gauss')

# Simulation:
data=foreach(n_sim=10^c(3:6,8),.combine='rbind')%do%{
	ptm=proc.time()
	n_binom=length(prob)
	set.seed(42)
	mat=matrix(rbinom(n_sim*n_binom,size,prob),nrow=n_binom,ncol=n_sim)
	
	S=colSums(mat)
	sim=sapply(X = 0:sum(size), FUN = function(x) {sum(S==x)/length(S)})
	
	print(proc.time()-ptm)
	
	data.table(s=0:sum(size),pdf=sim,type=n_sim)
	
}


data=rbind(data,gauss_approx,approx)
truth=data[type=='1e+08',]

merged=merge(truth[,list(s,pdf)],data,by='s',suffixes=c('_truth','_approx'))
merged=merged[type!='1e+08',]


p=ggplot(merged,aes(pdf_truth,pdf_approx))+
	geom_point()+
	facet_grid(~type)+
	geom_abline(intercept=0,slope=1)+
	theme_bw()+
	xlab('Truth')+
	ylab('Approximation')
save_plot('/Users/boshliu/Documents/tools/sinib/manuscript/figures/health_monitor_truth_vs_approx_pdf.pdf',p,base_height=2,base_width=8)

merged[,Error:=pdf_truth-pdf_approx]
merged[,`Relative Error`:=(pdf_truth-pdf_approx)/pdf_truth]
merged = melt(merged,measure.vars = c('Error','Relative Error'),variable.name = 'error_type',value.name = 'error')

p2=ggplot(merged,aes(s,error))+
	geom_point()+
	facet_grid(error_type~type,scales = 'free_y')+
	theme_bw()+
	xlab('Outcome')+
	ylab('Truth-Approx')+
	xlim(0,20) + 
	ylab(expression(frac(Truth - Approximation,Truth)~~~~~~~'Truth - Approximation'))
save_plot('/Users/boshliu/Documents/tools/sinib/manuscript/figures/health_monitor_truth_minus_approx_pdf.pdf',p2,base_height=4,base_width=8)



# time: 
ptm=proc.time()
n_binom=length(prob)
mat=matrix(rbinom(n_sim*n_binom,size,prob),nrow=n_binom,ncol=n_sim)
S=colSums(mat)
sim=sapply(X = 0:sum(size), FUN = function(x) {sum(S==x)/length(S)})
proc.time()-ptm

ptm=proc.time()
approx=dsinib(0:sum(size),size,prob)
proc.time()-ptm
